File: Solution/374.猜数字大小/374.猜数字大小.cpp
/*
 * @Descroption: LeetCode 374. 猜数字大小
 * @Author: EmoryHuang
 * @Date: 2021-05-28 15:47:10
 * @解题思路:
 * 二分法猜数字
 */

/**
 * Forward declaration of guess API.
 * @param  num   your guess
 * @return 	     -1 if num is lower than the guess number
 *			      1 if num is higher than the guess number
 *               otherwise return 0
 * int guess(int num);
 */

class Solution {
   public:
    int guessNumber(int n) {
        int l = 1, r = n;
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (guess(mid) <= 0)
                r = mid;
            else
                l = mid + 1;
        }
        return l;
    }
};

File: Solution/374.猜数字大小/374.猜数字大小.java
/*
 * @Descroption: LeetCode 374. 猜数字大小
 * @Author: EmoryHuang
 * @Date: 2023-07-20 10:28:15
 * @Method:
 * 二分查找
 */

import java.util.*;

/**
 * Forward declaration of guess API.
 * 
 * @param num your guess
 * @return -1 if num is higher than the picked number
 *         1 if num is lower than the picked number
 *         otherwise return 0
 *         int guess(int num);
 */

public class Solution extends GuessGame {
    public int guessNumber(int n) {
        int l = 1, r = n;
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (guess(mid) <= 0) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
        return l;
    }
}